common/src/chain/transaction/signature/sighash/mod.rs

```diff
@@ -40,7 +40,7 @@ fn hash_encoded_if_some<T: Encode>(val: &Option<T>, stream: &mut DefaultHashAlgo
 
 fn stream_signature_hash<T: Signable>(
     tx: &T,
-    inputs_utxos: &[&TxOutput],
+    inputs_utxos: &[Option<&TxOutput>],
```

Comment:
Ok. I might have gotten confused a bit. The `TxOuptut` here identifies what's being spent. In case of accounts, the account ID identifies what's being spent. I guess the account ID is already being committed to in the input, so that's fine.

However it also means that the signature commits to the account nonce for other inputs. I am not sure if that's the right thing to do, quite hard to work out all the implications. Does Ethereum do it? Or should the nonce be in witness? My instinct is that the current approach is correct because it basically emulates a sequence of single-input single-ouptut UTXOs but not 100% convinced. Thoughts? @TheQuantumPhysicist

Reply:
I can see an issue that a long living transaction won't be possible with this approach, but I don't see another issue with it. To be fair, ethereum has this issue too. The nonce can be used to invalidate or stall transactions.

Ethereum has one nonce per transaction, because the nonce binds to the "from" field, and the signature commits to all that.

While the nonce being in the witness sounds nice, I can see it introducing all kinds of complications. On one hand, the nonce is part of the signature, not part of the spendee; that's because, intiuitively, signing something should increment the nonce, in a perfect world. On the other hand, even if the nonce is in the witness, it should also be part of the witness commitment. So we'll have a circular problem where we have part of the witness data that should be "constrained" by the witness signature. What a mess!